packages/expo-permissions/android/src/main/java/expo/modules/permissions/PermissionsService.kt

```diff
@@ -1,65 +1,261 @@
 package expo.modules.permissions
 
+import android.Manifest
+import android.annotation.TargetApi
+import android.content.Context
+import android.content.Intent
 import android.content.pm.PackageManager
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.provider.Settings
 import android.support.v4.content.ContextCompat
+
 import com.facebook.react.modules.core.PermissionAwareActivity
 
+import expo.modules.permissions.requesters.*
+
 import org.unimodules.core.ModuleRegistry
 import org.unimodules.core.interfaces.ActivityProvider
 import org.unimodules.core.interfaces.InternalModule
+import org.unimodules.core.interfaces.LifecycleEventListener
+import org.unimodules.core.interfaces.services.UIManager
 import org.unimodules.interfaces.permissions.Permissions
 
+import kotlin.properties.Delegates
+
 private const val PERMISSIONS_REQUEST: Int = 13
 
-class PermissionsService: InternalModule, Permissions {
+internal const val EXPIRES_KEY = "expires"
+internal const val STATUS_KEY = "status"
+internal const val GRANTED_VALUE = "granted"
+internal const val DENIED_VALUE = "denied"
+internal const val UNDETERMINED_VALUE = "undetermined"
+internal const val ERROR_TAG = "E_PERMISSIONS"
+
+internal const val PERMISSION_EXPIRES_NEVER = "never"
+
+class PermissionsService(val context: Context): InternalModule, Permissions, LifecycleEventListener {
   private var mActivityProvider: ActivityProvider? = null
+  private var mRequesters: Map<String, PermissionRequester> by Delegates.notNull()
+
+  // state holders for asking for writing permissions
+  private var mWritingPermissionBeingAsked = false // change this directly before calling corresponding startActivity
+  private var mAskAsyncListener: Permissions.PermissionsRequesterListenerBundle? = null
+  private var mAskAsyncRequestedPermissionsTypes: Array<out String>? = null
+  private var mAskAsyncPermissionsTypesToBeAsked: ArrayList<String>? = null
```

Comment:
Again, think about this https://kotlinlang.org/docs/reference/properties.html#late-initialized-properties-and-variables

Reply:
These variables are used only when the user asks for system brightness permission. When we process this request, the system moves our app to the background so we need to save the current request. Next, we resume request with saved values. 
Because this situation can happen multiple times and we don't want to have values in these variables between requests, I think nullable types here are okay. 

Correct me if I'm wrong.